xml_struct_derive/src/serialize/codegen.rs

```diff
@@ -0,0 +1,588 @@
+            impl #impl_generics ::xml_struct::XmlSerializeAttr for #type_ident #ty_generics #where_clause {
+                fn serialize_as_attribute(&self, start_tag: &mut ::quick_xml::events::BytesStart, name: &str) {
+                    #body
+                }
+            }
+        }
+    } else {
+        TokenStream::default()
+    };
+
+    // Construct the final implementation from the type-specific sets of tokens.
+    quote! {
+        #[automatically_derived]
+        impl #impl_generics ::xml_struct::XmlSerialize for #type_ident #ty_generics #where_clause {
+            #as_element_impl
+
+            fn serialize_child_nodes<W: std::io::Write>(
+                &self,
+                writer: &mut ::quick_xml::writer::Writer<W>
+            ) -> Result<(), ::xml_struct::Error> {
+                #child_nodes_body
+
+                Ok(())
+            }
+        }
+
+        #attr_impl
+    }
+}
+
+/// The sets of tokens which make up the implementations or bodies of
+/// `XmlSerialize` and `XmlSerializeAttr` trait methods.
+pub(super) struct ImplTokenSets {
+    /// The implementation of `XmlSerialize::serialize_as_element()` if it is
+    /// necessary to override the provided default implementation.
+    as_element_impl: TokenStream,
+
+    /// The body of `XmlSerialize::serialize_child_nodes()`.
+    child_nodes_body: TokenStream,
+
+    /// The body of `XmlSerializeAttr::serialize_as_attribute()` if the type is
+    /// capable of being serialized as such.
+    as_attr_body: Option<TokenStream>,
+}
+
+/// Generates the sets of tokens necessary for serializing a struct with the
+/// provided fields.
+pub(super) fn with_struct_fields(
+    fields: Vec<Field>,
+) -> impl FnOnce(&[XmlAttribute]) -> ImplTokenSets {
+    move |namespace_attrs| {
+        let Fields {
+            attr_fields,
+            child_fields,
+        } = partition_fields(fields);
+
+        let content_call = if !child_fields.is_empty() {
+            Some(quote! {
+                <Self as XmlSerialize>::serialize_child_nodes(self, writer)?;
+            })
+        } else {
+            None
+        };
+
+        let impl_body =
+            generate_xml_tag_calls(quote!(name), namespace_attrs, &attr_fields, content_call);
+
+        ImplTokenSets {
+            as_element_impl: quote! {
+                fn serialize_as_element<W: std::io::Write>(
+                    &self,
+                    writer: &mut ::quick_xml::writer::Writer<W>,
+                    name: &str,
+                ) -> Result<(), ::xml_struct::Error> {
+                    #impl_body
+
+                    Ok(())
+                }
+            },
+            child_nodes_body: generate_child_node_calls_for_fields(child_fields),
+
+            // Only unit-only enums can be serialized as attribute values.
+            as_attr_body: None,
+        }
+    }
+}
+
+/// Generates the sets of tokens necessary for serializing a unit enum as text
```

Comment:
suggestion: consistency with Rust reference terminology

```suggestion
/// Generates the sets of tokens necessary for serializing a unit-only enum as text
```